sql/expression/function/strcmp.go

```diff
@@ -92,11 +88,21 @@ func (s *StrCmp) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 		return nil, nil
 	}
 
-	str2, err := sql.ConvertToString(expr2, sql.LongText)
-	if err != nil {
-		return nil, err
+	if sql.IsText(s.Left.Type()) && sql.IsText(s.Right.Type()) {
+		st1 := s.Left.Type().(sql.StringType)
+		st2 := s.Right.Type().(sql.StringType)
+
+		if st1.Collation() == st2.Collation() {
+			return st1.Compare(expr1, expr2)
+		} else {
+			return nil, sql.ErrCollationIllegalMix.New(st1.Collation(), st2.Collation())
+		}
+	} else if sql.IsText(s.Left.Type()) {
+		return s.Left.Type().Compare(expr1, expr2)
+	} else if sql.IsText(s.Right.Type()) {
+		return s.Right.Type().Compare(expr1, expr2)
+	} else {
+		st := sql.MustCreateString(sqltypes.Text, 10, ctx.GetCollation())
```

Comment:
This seems like a good spot to drop a link to MySQL reference documentation to explain why we use the current session's collation setting: 
https://dev.mysql.com/doc/refman/5.7/en/charset-collation-coercibility.html

It's often helpful to leave an excerpt from MySQL docs or a link to them to explain why we do something in the code that may not be obvious to future code readers.